### ISSUE TITLE
JS Pull Consumer Stalls

### ISSUE THREAD

**Issue by abalmos** [nats-io/nats.rs#860]
Make sure that these boxes are checked before submitting your issue -- thank you!

 - [x] Included below version and environment information
 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### NATS version (`grep 'name = "nats"' Cargo.lock -A 1`)
0.28.0 (from `main` as 0.28.0 has not been pushed to crates.io as of this report)

#### rustc version (`rustc --version` - we support Rust 1.41 and up)
rustc 1.66.1 (90743e729 2023-01-10)

#### OS/Container environment:
Fedora 37

#### Steps or code to reproduce the issue:

JetStream pull consumer suddenly stops all activity, but no errors are reported and everything remains stalled out indefinitely. After restarting the client, the consumer resumes from the stalled spot and continues on, until it stalls again.

To reproduce the issue, I created a stream on `demo.nats.io` called `js-stall-bug`. It listens to subject `js-stall-bug`. You can send test data with

```nats -s demo.nats.io pub js-stall-bug --count=100000000 --sleep=100ms "{{Count}} @ {{Time}}"```

The following code demonstrates the issue. I should note that the bug sometimes presents quickly, and other times takes hours, but it always does eventually happen.

```rust
use std::str::from_utf8;

use async_nats::jetstream::{self, consumer::AckPolicy, Message};
use color_eyre::eyre::{Report, Result};
use futures::StreamExt;

#[tokio::main]
async fn main() -> Result<(), async_nats::Error> {
    tracing_subscriber::fmt::init();

    let client = async_nats::ConnectOptions::new()
        .event_callback(|event| async move {
            match event {
                other => println!("EVENT: {}", other),
            }
        })
        .connect("demo.nats.io")
        .await
        .map_err(Report::msg)?;

    let jetstream = jetstream::new(client);

    let consumer = jetstream
        .get_stream("js-stall-bug")
        .await?
        .create_consumer(jetstream::consumer::pull::Config {
            durable_name: Some("js-stall-bug".into()),
            filter_subject: "js-stall-bug".into(),
            ack_policy: AckPolicy::All,
            max_ack_pending: 5_000,
            ..Default::default()
        })
        .await?;

    let mut i = 0;
    loop {
        i = i + 1;
        println!("[{}] Pulling...", i);

        let messages: Vec<Message> = consumer
            .batch()
            .max_messages(5_000)
            .expires(std::time::Duration::from_secs(1))
            .messages()
            .await
            .expect("Batch subscription failure.")
            .map(|msg| msg.expect("Unexpected operational error from NATS"))
            .collect()
            .await;

        for m in messages.iter() {
            println!("{}", from_utf8(&m.payload)?);
        }

        if let Some(m) = messages.last() {
            m.ack().await?;
        }
    }
}
```

#### Expected result:

The server to print the test messages until processing all 100000000 messages.

#### Actual result:

The service eventually stops while the `nats pub` command continues to publish messages. The stream grows, and the consumer sequence stays constant. There are no useful messages logged at the stall with `RUST_LOG=trace`. After the stall occurs, the server and client trade PING/PONGs on a regular interval. It is as if `async_nats` decides there is nothing more to do and simply stops pulling from the consumer.

Here is a snippet of the last test I ran. It was against demo.nats.io and took about 7 minutes and 20 seconds to occur.

```
[615] Pulling...
4390 @ 8:02PM
4391 @ 8:02PM
4392 @ 8:02PM
4393 @ 8:02PM
4394 @ 8:02PM
4395 @ 8:02PM
4396 @ 8:02PM
[616] Pulling...
2023-03-03T01:02:39.544363Z DEBUG async_nats: received PONG
2023-03-03T01:03:39.541672Z DEBUG async_nats: received PONG
2023-03-03T01:04:39.549764Z DEBUG async_nats: received PONG
2023-03-03T01:05:39.545310Z DEBUG async_nats: received PONG
2023-03-03T01:06:39.542932Z DEBUG async_nats: received PONG
2023-03-03T01:07:39.559286Z DEBUG async_nats: received PONG
```

**Comment by Jarema** [comment]
Can you share consumer info after it stalled?

I think youre reaching max ack pending here, as you ack only once at the end, and messages pile up for a pretty long time with the sleep.

**Comment by abalmos** [comment]
I have a test running again and waiting for a stall.

Could you elaborate on what you mean by reaching ack wait? The pull's expires is 1 second and the data comes in at about ~10 msg / second. I've never really see anything back up like that.

**Comment by abalmos** [comment]
Here is the consumer info after a stall

```
> nats -s demo.nats.io consumer info js-stall-bug
? Select a Consumer js-stall-bug
Information for Consumer js-stall-bug > js-stall-bug created 2023-03-02T19:35:26-05:00

Configuration:

                Name: js-stall-bug
           Pull Mode: true
      Filter Subject: js-stall-bug
      Deliver Policy: All
          Ack Policy: All
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 5,000
   Max Waiting Pulls: 512

State:

   Last Delivered Message: Consumer sequence: 24,803 Stream sequence: 24,796 Last delivery: 17m59s ago
     Acknowledgment floor: Consumer sequence: 24,797 Stream sequence: 24,790 Last Ack: 18m0s ago
         Outstanding Acks: 6 out of maximum 5,000
     Redelivered Messages: 0
     Unprocessed Messages: 6,942
            Waiting Pulls: 0 of maximum 512
```

I didn't notice the stall right away, so I let 6,942 extra messages build up. But otherwise, the info seems reasonable to me?

**Comment by Jarema** [comment]
I was able to reproduce the issue.
regarding the ack-wait - I missed the expire in config :).

**Comment by Jarema** [comment]
@abalmos it seems that for some reason, request is timed out but actual timeout is either not received from server, or not registered properly by client.

`stream` and `messages` methods, as long-running processes has protection against that, so until I fix it here, should be safe to use those two.

**Comment by Jarema** [comment]
Fix in nats-server has been merge https://github.com/nats-io/nats-server/pull/3942 

But additional check for timeout will be added to the client too.
Then, the issue will be closed.

**Comment by michaelmenon** [comment]
Just to confirm is this issue fixed in the  "0.28.0" version of the async-nats crate.

**Comment by abalmos** [comment]
@michaelmenon I _believe_ [this PR](https://github.com/nats-io/nats.rs/pull/862) fixes it in the client.

The server side fix is already merged into `nats-server` master, but I don't believe is released yet. I was hoping they would do a quick minor release as it is a pretty critical bug. Hopefully that can happen soon.

I understand that either one of the two fixes should be an effort to restore a stable consumer. I have a note on the PR to verify that thinking, but so far it has been stable in my testing.

**Comment by michaelmenon** [comment]
Ok thanks @abalmos for the info, hopefully `nats-server` is updated soon with the fix. :)

**Comment by Jarema** [comment]
#862 guards against server the server bug.

Closing this one. Release will happen shortly.
